core/src/main/scala/com/github/johnynek/paiges/Doc.scala

```diff
@@ -260,49 +291,140 @@ sealed abstract class Doc extends Product with Serializable {
    * are equivalent.
    */
   def compare(that: Doc): Int =
-    Doc.compareTree(Doc.toDocTree(this), Doc.toDocTree(that))
+    DocTree.compareTree(DocTree.toDocTree(this), DocTree.toDocTree(that))
+
+  /**
+   * Convert this Doc to a single-line representation.
+   *
+   * All newlines are replaced with spaces (and optional indentation
+   * is ignored). The resulting Doc will never render any newlines, no
+   * matter what width is used.
+   */
+  def flatten: Doc =
+    this match {
+      case Empty => Empty
+      case Line => Doc.space
+      case str@Text(_) => str
+      case Nest(i, d) => d.flatten // no Line, so Nest is irrelevant
+      case Concat(a, b) => Concat(a.flatten, b.flatten)
+      case Union(a, _) => a.flatten
+    }
+
+  /**
+   * This method is similar to flatten, but returns None if no
+   * flattening was needed (i.e. if no newlines were present).
+   *
+   * As with flatten, the resulting Doc (if any) will never render any
+   * newlines, no matter what width is used.
+   */
+  def flattenOption: Option[Doc] =
+    this match {
+      case Empty | Text(_) => None
+      case Line => Some(Doc.space)
```

Comment:
This is one of the main features we are thinking about how to add. We have to see how to do it, but I *think* it will be possible.